src/Concretize.hs

```diff
@@ -449,19 +449,31 @@ instantiateGenericSumtype typeEnv originalStructTy@(StructTy _ originalTyVars) g
                                             : concat okDeps
                     Left err -> Left err
 
+-- Resolves dependencies for sumtype cases.
+-- NOTE: This function only accepts cases that are in "canonical form"
+-- (Just [x]) aka XObj (Lst (Sym...) (Arr members))
+-- On other cases it will return an error.
 depsForCase :: TypeEnv -> XObj -> Either TypeError [XObj]
 depsForCase typeEnv x@(XObj (Lst [_, XObj (Arr members) _ _]) _ _) =
   concat <$>
     mapM (\m -> case xobjToTy m of
                Just okTy -> concretizeType typeEnv okTy
                Nothing -> error ("Failed to convert " ++ pretty m ++ " to a type: " ++ pretty x))
     members
+depsForCase _ x = Left (InvalidSumtypeCase x)
 
 replaceGenericTypeSymbolsOnCase :: Map.Map String Ty -> XObj -> XObj
 replaceGenericTypeSymbolsOnCase mappings singleCase@(XObj (Lst (caseName : caseMembers)) i t) =
   XObj (Lst (caseName : map replacer caseMembers)) i t
   where replacer memberXObj =
           replaceGenericTypeSymbols mappings memberXObj
+-- Handle cases like `(State a) Done (Value [a]))`
+-- `Done` is a Sym, not a Lst. DepsForCase, like this function
+-- expects and only matches on a Lst, so we convert the problematic cases to a
+-- canonical form. (see `depsForCase` above
+replaceGenericTypeSymbolsOnCase mappings nakedCase@(XObj (Sym (SymPath _ _) _) i t) =
+  XObj (Lst [nakedCase, XObj (Arr []) i t]) i t
```

Comment:
I know there is some other pass that does this transformation too. It obviously doesn't hurt to do it here too, since this must be earler in the process. But I might add a comment about that in the code too. Will merge though.

Reply:
Essentially the same thing happens in `SumtypeCases.hs` from what I can tell--I think that's what permits the initial definition to come through--but somehow the translation doesn't make it over to `Concretize`

Maybe we should abstract it out into a function or type synonym? 

`canoncialSumtypeCase` or `validSumtypeCase` or `concreteSumtypeCase` or something? `XObj -> XObj`